homeassistant/components/decora_wifi/light.py

```diff
@@ -17,88 +13,68 @@
     SUPPORT_TRANSITION,
     LightEntity,
 )
-from homeassistant.const import CONF_PASSWORD, CONF_USERNAME, EVENT_HOMEASSISTANT_STOP
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
 import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
+
+from .common import DecoraWifiEntity, DecoraWifiPlatform
+from .const import DOMAIN
 
 _LOGGER = logging.getLogger(__name__)
 
-# Validation of the user's configuration
+# Validation of the user's yaml configuration schema
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {vol.Required(CONF_USERNAME): cv.string, vol.Required(CONF_PASSWORD): cv.string}
+    {
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+    }
 )
 
-NOTIFICATION_ID = "leviton_notification"
-NOTIFICATION_TITLE = "myLeviton Decora Setup"
-
-
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up the Decora WiFi platform."""
-
-    email = config[CONF_USERNAME]
-    password = config[CONF_PASSWORD]
-    session = DecoraWiFiSession()
-
-    try:
-        success = session.login(email, password)
-
-        # If login failed, notify user.
-        if success is None:
-            msg = "Failed to log into myLeviton Services. Check credentials."
-            _LOGGER.error(msg)
-            hass.components.persistent_notification.create(
-                msg, title=NOTIFICATION_TITLE, notification_id=NOTIFICATION_ID
-            )
-            return False
-
-        # Gather all the available devices...
-        perms = session.user.get_residential_permissions()
-        all_switches = []
-        for permission in perms:
-            if permission.residentialAccountId is not None:
-                acct = ResidentialAccount(session, permission.residentialAccountId)
-                for residence in acct.get_residences():
-                    for switch in residence.get_iot_switches():
-                        all_switches.append(switch)
-            elif permission.residenceId is not None:
-                residence = Residence(session, permission.residenceId)
-                for switch in residence.get_iot_switches():
-                    all_switches.append(switch)
-
-        add_entities(DecoraWifiLight(sw) for sw in all_switches)
-    except ValueError:
-        _LOGGER.error("Failed to communicate with myLeviton Service")
-
-    # Listen for the stop event and log out.
-    def logout(event):
-        """Log out..."""
-        try:
-            if session is not None:
-                Person.logout(session)
-        except ValueError:
-            _LOGGER.error("Failed to log out of myLeviton Service")
-
-    hass.bus.listen(EVENT_HOMEASSISTANT_STOP, logout)
```

Comment:
Do we need to logout when Home Assistant is restarting?  The config entry doesn't get unloaded at shutdown (only reload)

Reply:
That would be good, yes. Logout only needs to be called once since all the entities share one API session, so I'm guessing this listener should now get set up where the session originates, in `async_setup_entry`.

Reply:
Hm, with 
```
    # Listen for the stop event and log out.
    hass.bus.listen(EVENT_HOMEASSISTANT_STOP, session.teardown)
```
added to `async_setup_entry`, `test_reauth_validate_invalid_login` is deadlocking, so either I need to do that somewhere else or my test has a bug, or my syntax is wrong.  I'll push the rest of the changes and come back to this one after I've slept.

Reply:
Ended up putting it in `DecoraWifiPlatform` setup and teardown.